ouroboros-network/src/Ouroboros/Network/Socket.hs

```diff
@@ -122,117 +123,210 @@ hexDump :: BL.ByteString -> String -> IO ()
 hexDump buf out | BL.empty == buf = say out
 hexDump buf out = hexDump (BL.tail buf) (out ++ printf "0x%02x " (BL.head buf))
 
+
 -- |
--- Run a node using @'NetworkInterface'@ using a socket.  It will start to
--- listen on incomming connections on the supplied @'nodeAddress'@, and returns
--- @'NetworkNode'@ which let one connect to other peers (by opening a new
--- TCP connection) or shut down the node.
---
--- When connecting to a remote node using @connectTo@ th
-withNetworkNode
-  :: forall ptcl t r.
+-- Connect to a remote node.  It is using bracket to enclose the underlying
+-- socket acquisition.  This implies that when the continuation exits the
+-- underlying bearer will get closed.
+connectTo
+  :: forall ptcl.
      ( Mx.ProtocolEnum ptcl
      , Ord ptcl
      , Enum ptcl
      , Bounded ptcl
      )
-  => NetworkInterface ptcl Socket.AddrInfo IO r
-  -> (NetworkNode Socket.AddrInfo IO r -> IO t)
-  -> IO t
-withNetworkNode NetworkInterface {nodeAddress, protocols} k =
-    bracket mkSocket Socket.close $ \sd -> do
+  => MuxApplication ClientApp ptcl IO
+  -- ^ application to run over the connection
+  -> Socket.AddrInfo
+  -- ^ address of the peer we want to connect to
+  -> IO ()
+connectTo app remoteAddr =
```

Comment:
The client address is still needed. With this change it is impossible to configure a node to only use a given address and/or port.

Reply:
Is this the right place to do that?  I think, It's the server responsibility to bind to a given address and/or port.

Reply:
We want both the client and the server to use the same port number and  the ability to restrict the client to only use a specific ip address. Without the ability to supply a local address and bind to it the address will depend on current routing information  and the port will be a free ephemeral port.

By supping 0 as client port address we can still keep the behavior you have now,  but it would be a configuration choice.

Reply:
ok, I'll add it.